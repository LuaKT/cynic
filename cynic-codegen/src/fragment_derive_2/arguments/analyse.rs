use super::parsing;

pub struct Output;

pub fn analyze(literals: Vec<parsing::FieldArgument>, schema: ()) -> Output {
    // Somehow implement this.
    // Though it's not that simple, as I really need more info than the schema parsing in this module
    // currently gives me.

    todo!("analyze not implemented")
}
