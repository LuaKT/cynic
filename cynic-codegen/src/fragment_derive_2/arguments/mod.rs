mod analyse;
mod parsing;

pub use parsing::{arguments_from_field_attrs, FieldArgument};
