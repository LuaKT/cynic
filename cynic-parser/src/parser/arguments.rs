use super::{named_type, selection_set, Parser};
use crate::{lexer::Token, syntax::SyntaxKind::*};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Context {
    Constant,
    NonConstant,
}

pub(super) fn arguments(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::OpenParen));
    parser.builder.start_node(ARGUMENTS.into());
    parser.bump();
    // TODO: almost wonder if bump should skip ws automatically...
    parser.skip_ws();

    loop {
        match parser.current() {
            None => {
                parser.error("expected arguments");
                break;
            }
            Some(Token::Name) => {
                argument(parser, context);
            }
            Some(Token::CloseParen) => {
                parser.bump();
                parser.skip_ws();
                break;
            }
            _ => {
                // TODO: Is this good?  no idea
                parser.error("expected a name or )");
                break;
            }
        }
    }

    parser.builder.finish_node();
}

fn argument(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(ARGUMENT.into());
    parser.bump();
    parser.skip_ws();

    match parser.current() {
        Some(Token::Colon) => {
            parser.bump();
        }
        _ => {
            parser.error("expected colon");
        }
    }

    value(parser, context);
    parser.skip_ws();
    parser.builder.finish_node();
}

pub(super) fn value(parser: &mut Parser, context: Context) {
    // TODO: Insert a value node here
    parser.builder.start_node(VALUE.into());
    parser.skip_ws();
    match parser.current() {
        Some(Token::Dollar) => {
            if context == Context::Constant {
                parser.error("Found variable in constant context");
                // Continue parsing anyway though...
            }
            super::variable(parser)
        }
        Some(Token::NegativeSign) | Some(Token::Number) => number(parser),
        Some(Token::Quote) | Some(Token::BlockQuote) => string(parser),
        Some(Token::Name) => {
            match parser.current_str().unwrap() {
                "true" | "false" => boolean(parser),
                "null" => null(parser),
                _ => enum_value(parser),
            }
            // Could be boolean, null, or an enum
        }
        Some(Token::OpenCurly) => object(parser, context),
        Some(Token::OpenSquare) => list(parser, context),
        _ => parser.error("expected a value"),
    }
    parser.skip_ws();
    parser.builder.finish_node();
}

fn number(parser: &mut Parser) {
    assert!(matches!(
        parser.current(),
        Some(Token::Number) | Some(Token::NegativeSign)
    ));
    let checkpoint = parser.builder.checkpoint();

    if let Some(Token::NegativeSign) = parser.current() {
        parser.bump();
    }

    match parser.current() {
        Some(Token::Number) => {
            parser.bump();
        }
        _ => parser.error("expected number"),
    }

    if let Some(Token::ExponentPart) | Some(Token::FractionalPart) = parser.current() {
        parser.builder.start_node_at(checkpoint, FLOAT_VALUE.into());
        if let Some(Token::FractionalPart) = parser.current() {
            parser.bump();
        }
        if let Some(Token::ExponentPart) = parser.current() {
            parser.bump();
        }
    } else {
        parser
            .builder
            .start_node_at(checkpoint, INTEGER_VALUE.into());
    }

    parser.builder.finish_node();
}

fn string(parser: &mut Parser) {
    assert!(matches!(
        parser.current(),
        Some(Token::Quote) | Some(Token::BlockQuote)
    ));
    // TODO: probably want to be able to differentiate between normal & block strings.
    parser.builder.start_node(STRING_VALUE.into());

    if parser.current() == Some(Token::Quote) {
        parser.bump();
        loop {
            match parser.current() {
                None => {
                    parser.error("unexpected end of input while parsing string");
                    break;
                }
                Some(Token::Quote) => {
                    parser.bump();
                    break;
                }
                Some(Token::LineTerminator) => {
                    parser.error("missing \"");
                    break;
                }
                _ => {
                    parser.bump_as(STRING_CONTENTS);
                }
            }
        }
    } else {
        parser.bump();
        loop {
            // TODO: I'm pretty sure this doesn't treat escape characters
            // correctly (there's also some mention of escaped block quotes that
            // I don't quite understand - figure this out later.)
            match parser.current() {
                None => {
                    parser.error("unexpected end of input while parsing string");
                    break;
                }
                Some(Token::BlockQuote) => {
                    parser.bump();
                    break;
                }
                _ => {
                    parser.bump_as(STRING_CONTENTS);
                }
            }
        }
    }

    parser.builder.finish_node();
}

fn object(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::OpenCurly));
    parser.builder.start_node(OBJECT_VALUE.into());
    parser.bump();
    parser.skip_ws();

    loop {
        match parser.current() {
            None => {
                parser.error("missing }");
            }
            Some(Token::CloseCurly) => {
                parser.bump();
                break;
            }
            Some(Token::Name) => {
                object_field(parser, context);
            }
            _ => {
                parser.error("expected name");
                break;
            }
        }
    }

    parser.builder.finish_node();
}

fn object_field(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(OBJECT_FIELD.into());
    parser.bump();
    parser.skip_ws();

    match parser.current() {
        Some(Token::Colon) => {
            parser.bump();
        }
        _ => parser.error("expected colon"),
    }
    value(parser, context);

    parser.builder.finish_node();
}

fn list(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::OpenSquare));
    parser.builder.start_node(LIST_VALUE.into());
    parser.bump();
    parser.skip_ws();
    loop {
        match parser.current() {
            None => {
                parser.error("missing ]");
                break;
            }
            Some(Token::CloseSquare) => {
                parser.bump();
                break;
            }
            _ => {}
        }
        value(parser, context);
        parser.skip_ws();
    }

    parser.builder.finish_node();
}

fn boolean(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));

    parser.builder.start_node(BOOL_VALUE.into());

    match parser.current_str() {
        Some("true") => parser.bump_as(TRUE),
        Some("false") => parser.bump_as(FALSE),
        _ => panic!("boolean pre-condition not met"),
    }

    parser.builder.finish_node();
}

fn null(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    assert_eq!(parser.current_str(), Some("null"));

    parser.bump_as(NULL);
}

fn enum_value(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));

    parser.builder.start_node(ENUM_VALUE.into());
    parser.bump();
    parser.builder.finish_node();
}
