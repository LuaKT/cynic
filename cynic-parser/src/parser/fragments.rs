use super::{arguments, directives::maybe_directives, named_type, selection_set, Parser};
use crate::{lexer::Token, syntax::SyntaxKind::*};

pub(super) fn fragment(parser: &mut Parser) {
    fragment_name(parser);

    type_condition(parser);

    maybe_directives(parser, arguments::Context::NonConstant);

    parser.skip_ws();
    match parser.current() {
        Some(Token::OpenCurly) => selection_set(parser),
        _ => parser.error("expected selection set"),
    }
}

pub(super) fn fragment_spread(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Spread));
    parser.builder.start_node(FRAGMENT_SPREAD.into());
    parser.bump();
    parser.skip_ws();
    if let Some(Token::Name) = parser.current() {
        fragment_name(parser);
    }

    maybe_directives(parser, arguments::Context::NonConstant);

    parser.skip_ws();
    if parser.current() == Some(Token::OpenCurly) {
        super::selection_set(parser);
    }

    parser.builder.finish_node();
}

fn fragment_name(parser: &mut Parser) {
    parser.skip_ws();
    match parser.current() {
        None => parser.error("expected fragment name"),
        Some(Token::Name) if parser.current_str() != Some("on") => {
            parser.builder.start_node(FRAGMENT_NAME.into());
            parser.bump();
            parser.builder.finish_node();
        }
        _ => {
            parser.error("expected fragment name");
        }
    }
}

pub(super) fn inline_fragment(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Spread));
    parser.builder.start_node(INLINE_FRAGMENT.into());
    parser.bump();
    parser.skip_ws();
    if let Some((Token::Name, "on")) = parser.current_pair() {
        type_condition(parser);
    }

    maybe_directives(parser, arguments::Context::NonConstant);

    parser.skip_ws();
    match parser.current() {
        Some(Token::OpenCurly) => selection_set(parser),
        _ => parser.error("expected selection set"),
    }

    parser.builder.finish_node();
}

fn type_condition(parser: &mut Parser) {
    parser.skip_ws();
    match parser.current_pair() {
        Some((Token::Name, "on")) => {
            parser.builder.start_node(TYPE_CONDITION.into());
            parser.bump_as(ON);
            parser.skip_ws();
            if let Some(Token::Name) = parser.current() {
                named_type(parser);
            } else {
                parser.error("expected named type");
            }
            parser.builder.finish_node();
        }
        _ => {
            parser.error("expected a type condition");
        }
    }
}
