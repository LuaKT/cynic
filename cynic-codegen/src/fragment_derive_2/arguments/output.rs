use proc_macro2::Span;
use quote::{quote, ToTokens, TokenStreamExt};
use syn::Token;

use super::analyse::{AnalysedArguments, ArgumentValue};

pub struct Output<'a> {
    pub(super) analysed: AnalysedArguments<'a>,
    pub(super) schema_module: syn::Path,
}

impl ToTokens for Output<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        if self.analysed.arguments.is_empty() {
            return;
        }

        let argument_module = self
            .analysed
            .schema_field
            .argument_module()
            .to_path(&self.schema_module);

        for arg in &self.analysed.arguments {
            let arg_marker = proc_macro2::Ident::from(arg.schema_field.marker_ident());
            let value = &arg.value;

            tokens.append_all(quote! {
                field_builder.argument::<#argument_module::#arg_marker>()
                #value;
            });
        }
    }
}

impl ToTokens for ArgumentValue<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match self {
            ArgumentValue::Object(_) => todo!("obj"),
            ArgumentValue::List(_) => todo!("list"),
            ArgumentValue::Literal(lit) => tokens.append_all(quote! {
                .literal(#lit)
            }),
            ArgumentValue::Bool(b) => {
                let lit = syn::LitBool::new(*b, Span::call_site());
                tokens.append_all(quote! {
                    .literal(#lit)
                });
            }
            ArgumentValue::Variable(_) => todo!("variable"),
            ArgumentValue::Some(_) => todo!("some"),
            ArgumentValue::Null => todo!("null"),
        }
    }
}
