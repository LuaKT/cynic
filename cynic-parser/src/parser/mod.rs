use logos::{Logos, Span};
use rowan::{GreenNode, GreenNodeBuilder};

use crate::{
    lexer::Token,
    syntax::{
        SyntaxKind::{self, *},
        SyntaxNode,
    },
};

mod arguments;
mod fragments;

use arguments::{arguments, constant_value};
use fragments::{fragment, fragment_spread, inline_fragment};

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
pub struct Parse {
    green_node: GreenNode,
    #[allow(unused)]
    errors: Vec<(String, Option<Span>)>,
}

pub fn parse(text: &str) -> Parse {
    let mut lexer = Token::lexer(text);
    let mut tokens = Vec::new();
    while let Some(token) = lexer.next() {
        tokens.push((token, lexer.slice(), lexer.span()));
    }
    tokens.reverse();

    Parser {
        tokens,
        builder: GreenNodeBuilder::new(),
        errors: Vec::new(),
    }
    .parse()
}

impl Parse {
    fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }
}

struct Parser<'source> {
    tokens: Vec<(Token, &'source str, Span)>,
    builder: GreenNodeBuilder<'static>,

    // TODO: Do I want spanned errors?
    errors: Vec<(String, Option<Span>)>,
}

impl<'source> Parser<'source> {
    fn parse(mut self) -> Parse {
        self.builder.start_node(ROOT.into());
        let mut current_error = false;
        loop {
            // Ok, so parsing out definitions here.
            match executable_def(&mut self) {
                Res::Eof => {
                    break;
                }
                Res::ExpectedExecutableDef => {
                    self.builder.start_node(ERROR.into());
                    if !current_error {
                        self.error("expected executable definition");
                        current_error = true;
                    }
                    self.bump();
                    // TODO: better nodes somehow
                    self.builder.finish_node();
                }
                Res::Ok => {
                    current_error = false;
                }
            }
        }
        self.skip_ws();
        self.builder.finish_node();

        Parse {
            green_node: self.builder.finish(),
            errors: self.errors,
        }
    }

    fn current(&self) -> Option<Token> {
        self.tokens.last().map(|(t, _, _)| *t)
    }

    fn peek_next_non_ws(&self) -> Option<Token> {
        let mut rev_iter = self.tokens.iter().rev();
        rev_iter.next(); // Skip the first one.
        rev_iter
            .skip_while(|(t, _, _)| *t == Token::Whitespace || *t == Token::LineTerminator)
            .next()
            .map(|(t, _, _)| *t)
    }

    fn peek_next_str_non_ws(&self) -> Option<&'source str> {
        let mut rev_iter = self.tokens.iter().rev();
        rev_iter.next(); // Skip the first one.
        rev_iter
            .skip_while(|(t, _, _)| *t == Token::Whitespace || *t == Token::LineTerminator)
            .next()
            .map(|(_, s, _)| *s)
    }

    fn current_str(&self) -> Option<&'source str> {
        self.tokens.last().map(|(_, s, _)| *s)
    }

    fn current_pair(&self) -> Option<(Token, &'source str)> {
        self.tokens.last().map(|(t, s, _)| (*t, *s))
    }

    fn current_span(&self) -> Option<Span> {
        self.tokens.last().map(|(_, _, span)| span.clone())
    }

    fn error(&mut self, err: impl Into<String>) {
        self.errors.push((err.into(), self.current_span()));
    }

    fn bump_as(&mut self, kind: SyntaxKind) {
        let (_, text, _) = self.tokens.pop().unwrap();
        self.builder.token(kind.into(), text)
    }

    fn bump(&mut self) {
        let (kind, text, _) = self.tokens.pop().unwrap();
        self.builder.token(SyntaxKind::from(kind).into(), text);
    }

    fn skip_ws(&mut self) {
        // TODO: This probably needs to know how to deal with comments as well
        while self.current() == Some(Token::Whitespace)
            || self.current() == Some(Token::LineTerminator)
            || self.current() == Some(Token::Comma)
        {
            self.bump();
        }
    }
}

enum Res {
    Eof,
    Ok,
    ExpectedExecutableDef,
}

fn executable_def(parser: &mut Parser) -> Res {
    parser.skip_ws();
    match parser.current_pair() {
        None => {
            // Return some EOF indicator
            return Res::Eof;
        }
        Some((Token::Name, "query")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(QUERY_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "mutation")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(MUTATION_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "subscription")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(SUBSCRIPTION_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::OpenCurly, _)) => {
            parser.builder.start_node(OPERATION_DEF.into());
            selection_set(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "fragment")) => {
            parser.builder.start_node(FRAGMENT_DEF.into());
            parser.bump_as(FRAGMENT_KEYWORD);
            fragment(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        other => {
            return Res::ExpectedExecutableDef;
        }
    }
}

fn operation(parser: &mut Parser) {
    parser.skip_ws();
    if let Some(Token::Name) = parser.current() {
        parser.bump();
    }

    parser.skip_ws();
    if let Some(Token::OpenParen) = parser.current() {
        variable_defs(parser);
    }

    // TODO: directives

    parser.skip_ws();
    if let Some(Token::OpenCurly) = parser.current() {
        selection_set(parser);
    } else {
        todo!()
        // TODO: Error
    }
}

fn variable_defs(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::OpenParen));
    parser.builder.start_node(VARIABLE_DEFS.into());
    parser.bump();
    loop {
        parser.skip_ws();
        match parser.current() {
            None => break,
            Some(Token::Dollar) => {
                parser.builder.start_node(VARIABLE_DEF.into());
                variable(parser);
                parser.skip_ws();
                match parser.current() {
                    Some(Token::Colon) => {
                        parser.bump();
                    }
                    _ => parser.error("expected :"),
                }
                type_(parser);
                parser.skip_ws();
                if let Some(Token::Equals) = parser.current() {
                    parser.builder.start_node(DEFAULT_VALUE.into());
                    parser.bump();
                    constant_value(parser);
                    parser.builder.finish_node();
                }
                parser.builder.finish_node();
                // todo: default values
            }
            Some(Token::CloseParen) => {
                parser.bump();
                break;
            }
            _ => {
                todo!()
                // TODO: error
            }
        }
    }
    parser.builder.finish_node();
}

fn variable(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Dollar));

    parser.builder.start_node(VARIABLE.into());
    parser.bump();
    match parser.current() {
        Some(Token::Name) => parser.bump(),
        _ => parser.error("expected name"),
    }
    parser.builder.finish_node();
}

fn type_(parser: &mut Parser) {
    parser.skip_ws();
    match parser.current() {
        Some(Token::Name) => {
            parser.builder.start_node(TYPE.into());
            named_type(parser);
            if let Some(Token::Bang) = parser.current() {
                parser.bump();
            }
            parser.builder.finish_node();
        }
        Some(Token::OpenSquare) => {
            parser.builder.start_node(TYPE.into());
            parser.bump();
            type_(parser);
            match parser.current() {
                Some(Token::CloseSquare) => {
                    parser.bump();
                }
                _ => {
                    parser.error("expected ]");
                }
            }
            parser.skip_ws();
            if let Some(Token::Bang) = parser.current() {
                parser.bump()
            }
            parser.builder.finish_node();
        }
        _ => {
            parser.error("expected name or [");
        }
    }
    parser.skip_ws();
}

fn named_type(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(NAMED_TYPE.into());
    parser.bump();
    parser.builder.finish_node();
}

fn selection_set(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::OpenCurly));

    parser.builder.start_node(SELECTION_SET.into());
    parser.bump();

    loop {
        parser.skip_ws();
        match parser.current() {
            None => {
                parser.error("expected selection");
                break;
            }
            Some(Token::Name) => {
                parser.builder.start_node(SELECTION.into());
                field_selection(parser);
                parser.builder.finish_node();
            }
            Some(Token::CloseCurly) => {
                parser.bump();
                break;
            }
            Some(Token::Spread) => {
                parser.builder.start_node(SELECTION.into());
                match parser.peek_next_str_non_ws() {
                    Some("on") => inline_fragment(parser),
                    _ => fragment_spread(parser),
                }
                parser.builder.finish_node();
            }
            _ => {
                // TODO: is this good? not sure it is..
                parser.error("expected selection");
                break;
            }
        }
    }

    parser.builder.finish_node();
}

fn field_selection(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(FIELD_SELECTION.into());
    if let Some(Token::Colon) = parser.peek_next_non_ws() {
        parser.builder.start_node(ALIAS.into());

        // Take the name & colon
        parser.bump();
        parser.skip_ws();
        parser.bump();
        parser.builder.finish_node();
    }

    parser.skip_ws();
    match parser.current() {
        Some(Token::Name) => {
            parser.bump();
        }
        _ => parser.error("expected name"),
    }

    parser.skip_ws();
    if let Some(Token::OpenParen) = parser.current() {
        arguments(parser);
    }

    parser.skip_ws();
    if let Some(Token::At) = parser.current() {
        todo!()
        // TODO: parse directives
    }

    parser.skip_ws();
    if let Some(Token::OpenCurly) = parser.current() {
        selection_set(parser)
    }

    parser.builder.finish_node();
}

#[cfg(test)]
mod tests {
    use std::{fs::File, io::Read};

    use rstest::rstest;

    use super::*;

    #[test]
    fn test_simple_parse() {
        let query = r#"
query {
    posts {
        title
        content
    }
}
"#;
        let result = parse(query);
        assert_eq!(result.errors, vec![]);
        assert_eq!(result.green_node.to_string(), query);
        insta::assert_debug_snapshot!(result.syntax());
    }

    #[rstest]
    #[case::fragment("tests/queries/fragment.graphql")]
    #[case::fragment_spread("tests/queries/fragment_spread.graphql")]
    #[case::inline_fragment("tests/queries/inline_fragment.graphql")]
    #[case::minimal("tests/queries/minimal.graphql")]
    #[case::minimal_mutation("tests/queries/minimal_mutation.graphql")]
    #[case::minimal_query("tests/queries/minimal_query.graphql")]
    #[case::named_query("tests/queries/named_query.graphql")]
    #[case::nested_selection("tests/queries/nested_selection.graphql")]
    #[case::query_aliases("tests/queries/query_aliases.graphql")]
    #[case::query_vars("tests/queries/query_vars.graphql")]
    #[case::query_arguments("tests/queries/query_arguments.graphql")]
    #[case::query_float_arguments("tests/queries/query_float_arguments.graphql")]
    #[case::query_list_argument("tests/queries/query_list_argument.graphql")]
    #[case::query_nameless_vars("tests/queries/query_nameless_vars.graphql")]
    // Commenting this one out because it requires comma ignores & comments
    //#[case::query_vars("tests/queries/query_nameless_vars_multiple_fields.graphql")]
    #[case::query_nameless_vars_multiple_fields_canonical(
        "tests/queries/query_nameless_vars_multiple_fields_canonical.graphql"
    )]
    #[case::query_object_argument("tests/queries/query_object_argument.graphql")]
    #[case::string_literal("tests/queries/string_literal.graphql")]
    #[case::triple_quoted_literal("tests/queries/triple_quoted_literal.graphql")]
    #[case::query_var_default_float("tests/queries/query_var_default_float.graphql")]
    #[case::query_var_default_list("tests/queries/query_var_default_list.graphql")]
    #[case::query_var_default_object("tests/queries/query_var_default_object.graphql")]
    #[case::query_var_default_string("tests/queries/query_var_default_string.graphql")]
    #[case::query_var_defaults("tests/queries/query_var_defaults.graphql")]
    fn test_query_file(#[case] file: String) {
        let mut query = String::new();
        File::open(file)
            .unwrap()
            .read_to_string(&mut query)
            .unwrap();

        let result = parse(&query);

        assert_eq!(result.errors, vec![]);
        assert_eq!(result.green_node.to_string(), query);
        // TODO: Determine the snapshot name from the file - to avoid ordering issues...
        insta::assert_debug_snapshot!(result.syntax());
    }
}
