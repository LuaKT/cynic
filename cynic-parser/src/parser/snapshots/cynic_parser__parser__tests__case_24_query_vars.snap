---
source: cynic-parser/src/parser/mod.rs
expression: result.syntax()

---
ROOT@0..332
  COMMA@0..1 ","
  COMMA@1..2 ","
  COMMA@2..3 ","
  COMMA@3..4 ","
  COMMA@4..5 ","
  COMMA@5..6 ","
  COMMA@6..7 ","
  COMMA@7..8 ","
  COMMA@8..9 ","
  COMMA@9..10 ","
  COMMA@10..11 ","
  COMMA@11..12 ","
  COMMA@12..13 ","
  COMMA@13..14 ","
  COMMA@14..15 ","
  COMMA@15..16 ","
  COMMA@16..17 ","
  LINE_TERMINATOR@17..18 "\n"
  OPERATION_DEF@18..331
    QUERY_KEYWORD@18..23 "query"
    WHITESPACE@23..24 " "
    COMMA@24..25 ","
    COMMA@25..26 ","
    COMMA@26..27 ","
    COMMA@27..28 ","
    COMMA@28..29 ","
    COMMA@29..30 ","
    COMMA@30..31 ","
    WHITESPACE@31..32 " "
    VARIABLE_DEFS@32..72
      OPEN_PAREN@32..33 "("
      VARIABLE_DEF@33..52
        VARIABLE@33..41
          DOLLAR@33..34 "$"
          NAME@34..41 "houseId"
        COLON@41..42 ":"
        WHITESPACE@42..43 " "
        TYPE@43..50
          NAMED_TYPE@43..49
            NAME@43..49 "String"
          BANG@49..50 "!"
        COMMA@50..51 ","
        WHITESPACE@51..52 " "
      VARIABLE_DEF@52..71
        VARIABLE@52..65
          DOLLAR@52..53 "$"
          NAME@53..65 "streetNumber"
        COLON@65..66 ":"
        WHITESPACE@66..67 " "
        TYPE@67..71
          NAMED_TYPE@67..70
            NAME@67..70 "Int"
          BANG@70..71 "!"
      CLOSE_PAREN@71..72 ")"
    WHITESPACE@72..73 " "
    COMMA@73..74 ","
    COMMA@74..75 ","
    COMMA@75..76 ","
    COMMA@76..77 ","
    COMMA@77..78 ","
    COMMA@78..79 ","
    COMMA@79..80 ","
    COMMA@80..81 ","
    COMMA@81..82 ","
    COMMA@82..83 ","
    COMMA@83..84 ","
    COMMA@84..85 ","
    WHITESPACE@85..86 " "
    SELECTION_SET@86..331
      OPEN_CURLY@86..87 "{"
      WHITESPACE@87..88 " "
      COMMENT@88..98
        COMMENT_CONTENTS@88..89 "#"
        COMMENT_CONTENTS@89..90 " "
        COMMENT_CONTENTS@90..97 "comment"
        LINE_TERMINATOR@97..98 "\n"
      COMMA@98..99 ","
      COMMA@99..100 ","
      COMMA@100..101 ","
      COMMA@101..102 ","
      COMMA@102..103 ","
      COMMA@103..104 ","
      COMMA@104..105 ","
      COMMA@105..106 ","
      COMMA@106..107 ","
      COMMA@107..108 ","
      COMMA@108..109 ","
      COMMA@109..110 ","
      COMMA@110..111 ","
      COMMA@111..112 ","
      COMMA@112..113 ","
      COMMA@113..114 ","
      COMMA@114..115 ","
      COMMA@115..116 ","
      WHITESPACE@116..117 " "
      COMMENT@117..141
        COMMENT_CONTENTS@117..118 "#"
        COMMENT_CONTENTS@118..119 " "
        COMMENT_CONTENTS@119..125 "commas"
        COMMENT_CONTENTS@125..126 " "
        COMMENT_CONTENTS@126..132 "should"
        COMMENT_CONTENTS@132..133 " "
        COMMENT_CONTENTS@133..135 "be"
        COMMENT_CONTENTS@135..136 " "
        COMMENT_CONTENTS@136..140 "fine"
        LINE_TERMINATOR@140..141 "\n"
      WHITESPACE@141..143 "  "
      SELECTION@143..200
        FIELD_SELECTION@143..200
          NAME@143..148 "house"
          ARGUMENTS@148..163
            OPEN_PAREN@148..149 "("
            ARGUMENT@149..161
              NAME@149..151 "id"
              COLON@151..152 ":"
              VALUE@152..161
                WHITESPACE@152..153 " "
                VARIABLE@153..161
                  DOLLAR@153..154 "$"
                  NAME@154..161 "houseId"
            CLOSE_PAREN@161..162 ")"
            WHITESPACE@162..163 " "
          SELECTION_SET@163..200
            OPEN_CURLY@163..164 "{"
            LINE_TERMINATOR@164..165 "\n"
            WHITESPACE@165..169 "    "
            SELECTION@169..176
              FIELD_SELECTION@169..176
                NAME@169..171 "id"
                LINE_TERMINATOR@171..172 "\n"
                WHITESPACE@172..176 "    "
            SELECTION@176..185
              FIELD_SELECTION@176..185
                NAME@176..180 "name"
                LINE_TERMINATOR@180..181 "\n"
                WHITESPACE@181..185 "    "
            SELECTION@185..193
              FIELD_SELECTION@185..193
                NAME@185..188 "lat"
                LINE_TERMINATOR@188..189 "\n"
                WHITESPACE@189..193 "    "
            SELECTION@193..199
              FIELD_SELECTION@193..199
                NAME@193..196 "lng"
                LINE_TERMINATOR@196..197 "\n"
                WHITESPACE@197..199 "  "
            CLOSE_CURLY@199..200 "}"
      LINE_TERMINATOR@200..201 "\n"
      WHITESPACE@201..203 "  "
      SELECTION@203..265
        FIELD_SELECTION@203..265
          NAME@203..209 "street"
          ARGUMENTS@209..233
            OPEN_PAREN@209..210 "("
            ARGUMENT@210..231
              NAME@210..216 "number"
              COLON@216..217 ":"
              VALUE@217..231
                WHITESPACE@217..218 " "
                VARIABLE@218..231
                  DOLLAR@218..219 "$"
                  NAME@219..231 "streetNumber"
            CLOSE_PAREN@231..232 ")"
            WHITESPACE@232..233 " "
          SELECTION_SET@233..265
            OPEN_CURLY@233..234 "{"
            WHITESPACE@234..235 " "
            COMMENT@235..255
              COMMENT_CONTENTS@235..236 "#"
              COMMENT_CONTENTS@236..237 " "
              COMMENT_CONTENTS@237..241 "this"
              COMMENT_CONTENTS@241..242 " "
              COMMENT_CONTENTS@242..244 "is"
              COMMENT_CONTENTS@244..245 " "
              COMMENT_CONTENTS@245..246 "a"
              COMMENT_CONTENTS@246..247 " "
              COMMENT_CONTENTS@247..254 "comment"
              LINE_TERMINATOR@254..255 "\n"
            WHITESPACE@255..259 "    "
            SELECTION@259..264
              FIELD_SELECTION@259..264
                NAME@259..261 "id"
                LINE_TERMINATOR@261..262 "\n"
                WHITESPACE@262..264 "  "
            CLOSE_CURLY@264..265 "}"
      LINE_TERMINATOR@265..266 "\n"
      WHITESPACE@266..268 "  "
      SELECTION@268..329
        FIELD_SELECTION@268..329
          NAME@268..279 "houseStreet"
          ARGUMENTS@279..317
            OPEN_PAREN@279..280 "("
            ARGUMENT@280..294
              NAME@280..282 "id"
              COLON@282..283 ":"
              VALUE@283..294
                WHITESPACE@283..284 " "
                VARIABLE@284..292
                  DOLLAR@284..285 "$"
                  NAME@285..292 "houseId"
                COMMA@292..293 ","
                WHITESPACE@293..294 " "
            ARGUMENT@294..315
              NAME@294..300 "number"
              COLON@300..301 ":"
              VALUE@301..315
                WHITESPACE@301..302 " "
                VARIABLE@302..315
                  DOLLAR@302..303 "$"
                  NAME@303..315 "streetNumber"
            CLOSE_PAREN@315..316 ")"
            WHITESPACE@316..317 " "
          SELECTION_SET@317..329
            OPEN_CURLY@317..318 "{"
            LINE_TERMINATOR@318..319 "\n"
            WHITESPACE@319..323 "    "
            SELECTION@323..328
              FIELD_SELECTION@323..328
                NAME@323..325 "id"
                LINE_TERMINATOR@325..326 "\n"
                WHITESPACE@326..328 "  "
            CLOSE_CURLY@328..329 "}"
      LINE_TERMINATOR@329..330 "\n"
      CLOSE_CURLY@330..331 "}"
  LINE_TERMINATOR@331..332 "\n"

